File: 2023-02-20/sums.c
/* C solution to David Amos's coding challenge "Adding it all up":
 * https://discourse.davidamos.dev/t/adding-it-all-up/139 */

#include<stdio.h>
#include<time.h>

#define REPEATS 1000000

long loop(long n) {
    long ret_val = 0;
    if (n > 0) {
        for (long k = 0; k < n + 1; k++) {
            ret_val += k;
        }
        return ret_val;
    } else if (n == 0) {
        return 0;
    } else {
        return loop(-n);
    }
}

int main() {
    printf("%ld, %d \n", loop(12345), 12345 * 12346 / 2);
    long inputs[] = {14, -15, 120, 0};
    long starts[] = {-3, 0, -20};
    long ends[] = {3, 8, 20};
    clock_t t;
    float comp_time;
    for(int j = 0; j < 4; j++) {
        t = clock();
        for(int w = 0; w < REPEATS; w++) {
            loop(inputs[j]);
        }
        t = clock() - t;
        comp_time = ((float) (1000000000 * t))/(CLOCKS_PER_SEC * REPEATS);
        printf("n = %ld: %0.3f ns\n", inputs[j], comp_time);
    }
    for(int j = 0; j < 3; j++) {
        t = clock();
        for(int w = 0; w < REPEATS; w++) {
            long outval = 0;
            for(long u = starts[j]; u < ends[j]; u++){
                outval += loop(u);
            }
        }
        t = clock() - t;
        comp_time = ((float) (1000000000 * t))/(CLOCKS_PER_SEC * REPEATS);
        printf("range %ld to %ld: %0.3f ns\n", starts[j], ends[j], comp_time);
    }
}

File: 2023-02-20/sums.py
"""
Several Python solutions to David Amos's Weekly Coding Challenge:
    https://discourse.davidamos.dev/t/adding-it-all-up/139
"""

_cache: dict[int, int] = {}  # Could've use functools.cache decorator


def loop(n: int) -> int:
    """Return 1 + 2 + ... + n if n > 0, 0 if n == 0, and -1 - 2 - ... + n if n < 0."""
    if not isinstance(n, int):
        raise TypeError(f"Arugment {n!r} is not an integer.")
    if n > 0:
        ret_val = 0
        for k in range(1, n + 1):
            ret_val += k
        return ret_val
    elif n == 0:
        return 0
    else:
        return -loop(-n)


def listcomp(n: int) -> int:
    """Return 1 + 2 + ... + n if n > 0, 0 if n == 0, and -1 - 2 - ... + n if n < 0."""
    if not isinstance(n, int):
        raise TypeError(f"Arugment {n!r} is not an integer.")
    if n > 0:
        return sum([k for k in range(n + 1)])
    elif n == 0:
        return 0
    else:
        return -listcomp(-n)


def recursive(n: int) -> int:
    """Return 1 + 2 + ... + n if n > 0, 0 if n == 0, and -1 - 2 - ... + n if n < 0."""
    if not isinstance(n, int):
        raise TypeError(f"Arugment {n!r} is not an integer.")
    if n == 0:
        return 0
    elif n > 0:
        return n + recursive(n - 1)
    else:
        return -recursive(-n)


def cached_recursive(n: int) -> int:
    """Return 1 + 2 + ... + n if n > 0, 0 if n == 0, and -1 - 2 - ... + n if n < 0."""
    if not isinstance(n, int):
        raise TypeError(f"Arugment {n!r} is not an integer.")
    if n not in _cache:
        if n == 0:
            _cache[n] = 0
        elif n > 0:
            _cache[n] = n + recursive(n - 1)
        else:
            _cache[n] = -recursive(-n)
    return _cache[n]


def gauss_sum(n: int) -> int:
    """Return 1 + 2 + ... + n if n > 0, 0 if n == 0, and -1 - 2 - ... + n if n < 0."""
    if not isinstance(n, int):
        raise TypeError(f"Arugment {n!r} is not an integer.")
    return n * (abs(n) + 1) // 2
